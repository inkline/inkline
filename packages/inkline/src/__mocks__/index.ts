export * from '@inkline/inkline/__mocks__/globalPropertiesMock';
export * from '@inkline/inkline/__mocks__/localStorageMock';
export * from '@inkline/inkline/__mocks__/matchMediaMock';
