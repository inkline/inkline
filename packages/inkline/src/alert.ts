export * from '@inkline/component-alert';
