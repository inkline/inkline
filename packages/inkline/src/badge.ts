export * from '@inkline/component-badge';