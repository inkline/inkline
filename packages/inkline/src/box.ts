export { Box, GridBox } from '@inkline/component-box';
export * from '@inkline/component-box/theme';
