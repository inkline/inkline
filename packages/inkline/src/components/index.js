export { default as IAlert } from '@inkline/inkline/src/components/IAlert';
export { default as IBadge } from '@inkline/inkline/src/components/IBadge';
export { default as IBreadcrumb } from '@inkline/inkline/src/components/IBreadcrumb';
export { default as IBreadcrumbItem } from '@inkline/inkline/src/components/IBreadcrumbItem';
export { default as IButton } from '@inkline/inkline/src/components/IButton';
export { default as IButtonGroup } from '@inkline/inkline/src/components/IButtonGroup';
export { default as ICard } from '@inkline/inkline/src/components/ICard';
export { default as ICollapsible } from '@inkline/inkline/src/components/ICollapsible';
export { default as ICollapsibleItem } from '@inkline/inkline/src/components/ICollapsibleItem';
export { default as IColumn } from '@inkline/inkline/src/components/IColumn';
export { default as IContainer } from '@inkline/inkline/src/components/IContainer';
export { default as IDatatable } from '@inkline/inkline/src/components/IDatatable';
export { default as IDropdown } from '@inkline/inkline/src/components/IDropdown';
export { default as IDropdownDivider } from '@inkline/inkline/src/components/IDropdownDivider';
export { default as IDropdownItem } from '@inkline/inkline/src/components/IDropdownItem';
export { default as IDropdownMenu } from '@inkline/inkline/src/components/IDropdownMenu';
export { default as IHamburgerMenu } from '@inkline/inkline/src/components/IHamburgerMenu';
export { default as IHeader } from '@inkline/inkline/src/components/IHeader';
export { default as IIcon } from '@inkline/inkline/src/components/IIcon';
export { default as ILayout } from '@inkline/inkline/src/components/ILayout';
export { default as ILayoutAside } from '@inkline/inkline/src/components/ILayoutAside';
export { default as ILayoutContent } from '@inkline/inkline/src/components/ILayoutContent';
export { default as ILayoutFooter } from '@inkline/inkline/src/components/ILayoutFooter';
export { default as ILayoutHeader } from '@inkline/inkline/src/components/ILayoutHeader';
export { default as ILinkable } from '@inkline/inkline/src/components/ILinkable';
export { default as IListGroup } from '@inkline/inkline/src/components/IListGroup';
export { default as IListGroupItem } from '@inkline/inkline/src/components/IListGroupItem';
export { default as ILoader } from '@inkline/inkline/src/components/ILoader';
export { default as IMedia } from '@inkline/inkline/src/components/IMedia';
export { default as IModal } from '@inkline/inkline/src/components/IModal';
export { default as INav } from '@inkline/inkline/src/components/INav';
export { default as INavItem } from '@inkline/inkline/src/components/INavItem';
export { default as INavbar } from '@inkline/inkline/src/components/INavbar';
export { default as INavbarBrand } from '@inkline/inkline/src/components/INavbarBrand';
export { default as INavbarItems } from '@inkline/inkline/src/components/INavbarItems';
export { default as IRow } from '@inkline/inkline/src/components/IRow';
export { default as IPagination } from '@inkline/inkline/src/components/IPagination';
export { default as IPopover } from '@inkline/inkline/src/components/IPopover';
export { default as IProgress } from '@inkline/inkline/src/components/IProgress';
export { default as IProgressBar } from '@inkline/inkline/src/components/IProgressBar';
export { default as ISidebar } from '@inkline/inkline/src/components/ISidebar';
export { default as ITable } from '@inkline/inkline/src/components/ITable';
export { default as ITabs } from '@inkline/inkline/src/components/ITabs';
export { default as ITab } from '@inkline/inkline/src/components/ITab';
export { default as ITooltip } from '@inkline/inkline/src/components/ITooltip';

export { default as ICheckable } from '@inkline/inkline/src/components/ICheckable';
export { default as ICheckableGroup } from '@inkline/inkline/src/components/ICheckableGroup';
export { default as ICheckbox } from '@inkline/inkline/src/components/ICheckbox';
export { default as ICheckboxButton } from '@inkline/inkline/src/components/ICheckboxButton';
export { default as ICheckboxGroup } from '@inkline/inkline/src/components/ICheckboxGroup';
export { default as ICheckboxButtonGroup } from '@inkline/inkline/src/components/ICheckboxButtonGroup';
export { default as IInput } from '@inkline/inkline/src/components/IInput';
export { default as IForm } from '@inkline/inkline/src/components/IForm';
export { default as IFormGroup } from '@inkline/inkline/src/components/IFormGroup';
export { default as IFormLabel } from '@inkline/inkline/src/components/IFormLabel';
export { default as IFormError } from '@inkline/inkline/src/components/IFormError';
export { default as IInputNumber } from '@inkline/inkline/src/components/IInputNumber';
export { default as IRadio } from '@inkline/inkline/src/components/IRadio';
export { default as IRadioButton } from '@inkline/inkline/src/components/IRadioButton';
export { default as IRadioGroup } from '@inkline/inkline/src/components/IRadioGroup';
export { default as IRadioButtonGroup } from '@inkline/inkline/src/components/IRadioButtonGroup';
export { default as ISelect } from '@inkline/inkline/src/components/ISelect';
export { default as ISelectOption } from '@inkline/inkline/src/components/ISelectOption';
export { default as ITextarea } from '@inkline/inkline/src/components/ITextarea';
export { default as IToggle } from '@inkline/inkline/src/components/IToggle';
