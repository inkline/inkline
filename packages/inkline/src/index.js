export * from '@inkline/inkline/src/plugin';
export * from '@inkline/inkline/src/components';
