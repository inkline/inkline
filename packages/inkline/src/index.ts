export * from '@inkline/i18n';
export * from '@inkline/utils';
export * from '@inkline/validation';
export * from '@inkline/vue';

export * from './components';
