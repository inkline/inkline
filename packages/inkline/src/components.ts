import type { Component } from 'vue';
import { Alert } from './alert';
import { Badge } from './badge';
import { Breadcrumb, BreadcrumbItem } from './breadcrumb';
import { Button } from './button';
import { ButtonGroup } from './button-group';
import { Card } from './card';
import { Checkbox, CheckboxGroup } from './checkbox';
import { Drawer } from './drawer';
import { FormGroup } from './form-group';
import { Expand } from './expand';
import { Grid } from './grid';
import { Icon } from './icon';
import { Input, Textarea } from './input';
import { Loader } from './loader';
import { Media } from './media';
import { Modal, ModalContainer } from './modal';
import { Nav, NavItem } from './nav';
import { Popover } from './popover';
import { Popup } from './popup';
import { Radio, RadioGroup } from './radio';
import { Toast, ToastContainer } from './toast';
import { Toggle } from './toggle';
import { Tooltip } from './tooltip';
import { Blockquote, Typography } from './typography';

export const components: Record<string, Component> = {
    Alert,
    Badge,
    Blockquote,
    Breadcrumb,
    BreadcrumbItem,
    Button,
    ButtonGroup,
    Card,
    Checkbox,
    CheckboxGroup,
    Drawer,
    FormGroup,
    Expand,
    Grid,
    Icon,
    Input,
    Loader,
    Media,
    Modal,
    ModalContainer,
    Nav,
    NavItem,
    Popover,
    Popup,
    Radio,
    RadioGroup,
    Textarea,
    Toast,
    ToastContainer,
    Toggle,
    Tooltip,
    Typography
};
