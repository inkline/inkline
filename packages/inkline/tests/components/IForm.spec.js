import { shallowMount } from '@vue/test-utils';
import { IForm } from '@inkline/inkline/src/components';

describe('Components', () => {
    describe('IForm', () => {
        let wrapper;

        const $inkline = {
            config: {
                validation: {
                    on: ['input']
                }
            }
        };

        beforeEach(() => {
            wrapper = shallowMount(IForm, {
                methods: {
                    created: IForm.created,
                    provide: IForm.provide
                },
                mocks: {
                    $inkline
                },
            });
        });

        it('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('inline', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.inline).toBeDefined();
                    expect(wrapper.vm.inline).toEqual(false);
                });
            });

            describe('value', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.value).toBeDefined();
                    expect(wrapper.vm.value).toEqual(null);
                });
            });
        });

        describe('provide()', () => {
            it('should return object containing parentForm as current instance', () => {
                expect(wrapper.vm.provide()).toEqual({ parentForm: wrapper.vm });
            });
        });

        describe('methods', () => {
            let inputWrapper;

            beforeEach(() => {
                jest.clearAllMocks();

                wrapper.setProps({
                    value: {
                        input: {
                            validateOn: 'input',
                            touch: () => ({}),
                            validate: () => ({ errors: {}, valid: true })
                        },
                        group: {
                            input: {}
                        }
                    }
                });

                inputWrapper = shallowMount({
                    data () {
                        return {
                            name: 'input',
                            schema: wrapper.vm.value.input
                        }
                    },
                    render () {},
                });
            });

            describe('getSchema()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.getSchema).toBeDefined();
                });

                it('should return form schema', () => {
                    const schema = {
                        input: {}
                    };

                    wrapper.setProps({ value: schema });

                    expect(wrapper.vm.getSchema()).toEqual(schema);
                });
            });

            describe('onFormControlInput()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.onFormControlInput).toBeDefined();
                });

                it('should return a function', () => {
                    expect(wrapper.vm.onFormControlInput(inputWrapper.vm)).toEqual(expect.any(Function));
                });

                it('should set form control schema value', () => {
                    const callback = wrapper.vm.onFormControlInput(inputWrapper.vm);
                    const value = 'abc';

                    callback(value);

                    expect(inputWrapper.vm.schema.value).toEqual(value);
                    expect(wrapper.vm.value.input.value).toEqual('abc');
                });

                it('should emit input event', () => {
                    const callback = wrapper.vm.onFormControlInput(inputWrapper.vm);
                    const spy = jest.spyOn(wrapper.vm, '$emit');

                    callback('abc');

                    expect(spy).toHaveBeenCalledWith('input', wrapper.vm.value);
                });
            });

            describe('onFormControlBlur()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.onFormControlBlur).toBeDefined();
                });

                it('should return a function', () => {
                    expect(wrapper.vm.onFormControlBlur(inputWrapper.vm)).toEqual(expect.any(Function));
                });

                it('should emit input event', () => {
                    const callback = wrapper.vm.onFormControlBlur(inputWrapper.vm);
                    const spy = jest.spyOn(inputWrapper.vm.schema, 'touch');

                    callback('abc');

                    expect(spy).toHaveBeenCalledWith({ getSchema: wrapper.vm.getSchema });
                });
            });

            describe('onFormControlValidate()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.onFormControlValidate).toBeDefined();
                });

                it('should return a function with existing event', () => {
                    expect(wrapper.vm.onFormControlValidate(inputWrapper.vm, 'input')).toEqual(expect.any(Function));
                });

                it('should return a function custom event', () => {
                    expect(wrapper.vm.onFormControlValidate(inputWrapper.vm, 'custom')).toEqual(expect.any(Function));
                });

                it('should call form control validate function', () => {
                    const callback = wrapper.vm.onFormControlValidate(inputWrapper.vm);
                    const spy = jest.spyOn(inputWrapper.vm.schema, 'validate');
                    const value = 'abc';

                    callback(value);

                    expect(spy).toHaveBeenCalledWith(value, { getSchema: wrapper.vm.getSchema });
                });

                it('should emit validate event', () => {
                    const callback = wrapper.vm.onFormControlValidate(inputWrapper.vm);
                    const spy = jest.spyOn(wrapper.vm, '$emit');

                    callback('abc');

                    expect(spy).toHaveBeenCalledWith('validate', wrapper.vm.value);
                });
            });

            describe('getFormControlValidationEvents()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.getFormControlValidationEvents).toBeDefined();
                });

                it('should return a form control validation event as array', () => {
                    const validateOn = 'input';

                    inputWrapper.setProps({
                        schema: {
                            validateOn
                        }
                    });

                    const events = wrapper.vm.getFormControlValidationEvents(inputWrapper.vm);

                    expect(events).toEqual([validateOn]);
                });

                it('should return a form control\'s multiple validation events as array', () => {
                    const validateOn = ['input', 'blur'];

                    inputWrapper.setData({
                        schema: {
                            validateOn
                        }
                    });

                    const events = wrapper.vm.getFormControlValidationEvents(inputWrapper.vm);

                    expect(events).toEqual(validateOn);
                });

                it('should return default validation event if validateOn not specified', (done) => {
                    inputWrapper.setData({
                        schema: {
                            validateOn: null
                        }
                    });

                    inputWrapper.vm.$inkline = {
                        config: {
                            validation: {
                                validateOn: ['input']
                            }
                        }
                    };

                    inputWrapper.vm.$nextTick(() => {
                        const events = wrapper.vm.getFormControlValidationEvents(inputWrapper.vm);

                        expect(events).toEqual(wrapper.vm.$inkline.config.validation.validateOn);

                        done();
                    });
                });
            });

            describe('add()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.add).toBeDefined();
                });

                it('should add "input" event listener to form control', () => {
                    const formSpy = jest.spyOn(wrapper.vm, 'onFormControlInput');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$on');

                    wrapper.vm.add(inputWrapper.vm);

                    expect(formSpy).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(1, 'input', expect.any(Function));
                });

                it('should add "blur" event listener to form control', () => {
                    const formSpy = jest.spyOn(wrapper.vm, 'onFormControlBlur');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$on');

                    wrapper.vm.add(inputWrapper.vm);

                    expect(formSpy).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(2, 'blur', expect.any(Function));
                });

                it('should add validation event listener to form control', () => {
                    const formSpy1 = jest.spyOn(wrapper.vm, 'getFormControlValidationEvents');
                    const formSpy2 = jest.spyOn(wrapper.vm, 'onFormControlValidate');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$on');

                    wrapper.vm.add(inputWrapper.vm);

                    expect(formSpy1).toHaveBeenCalled();
                    expect(formSpy2).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(1, 'input', expect.any(Function));
                });


                it('should set the all schema tree entries as touched on "blur"', () => {
                    const spy = jest.spyOn(inputWrapper.vm.schema, 'touch');

                    wrapper.vm.add(inputWrapper.vm);
                    inputWrapper.vm.$emit('blur');

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith({ getSchema: wrapper.vm.getSchema });
                });

                it('should add the input schema\'s validateOn event listener to input', () => {
                    const spy = jest.spyOn(inputWrapper.vm, '$on');

                    wrapper.vm.add(inputWrapper.vm);

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith('input', expect.any(Function));
                });

                it('should add the input schema\'s multiple validateOn event listeners to input', () => {
                    const spy = jest.spyOn(inputWrapper.vm, '$on');

                    inputWrapper.vm.schema.validateOn = ['input', 'blur'];
                    wrapper.vm.add(inputWrapper.vm);

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith('input', expect.any(Function));
                    expect(spy).toHaveBeenCalledWith('blur', expect.any(Function));
                });

                it('should add the input schema\'s custom event listeners to input', () => {
                    const spy = jest.spyOn(inputWrapper.vm, '$on');

                    inputWrapper.vm.schema.validateOn = ['custom'];
                    wrapper.vm.add(inputWrapper.vm);

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith('custom', expect.any(Function));
                });

                it('should call inputSchema validate() on validateOn event', () => {
                    const spy = jest.spyOn(inputWrapper.vm.schema, 'validate');

                    wrapper.vm.add(inputWrapper.vm);
                    inputWrapper.vm.$emit('input', 10);

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith(10, { getSchema: wrapper.vm.getSchema });
                });
            });

            describe('remove()', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.remove).toBeDefined();
                });

                it('should remove "input" event listener to form control', () => {
                    const formSpy = jest.spyOn(wrapper.vm, 'onFormControlInput');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$off');

                    wrapper.vm.remove(inputWrapper.vm);

                    expect(formSpy).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(1, 'input', expect.any(Function));
                });

                it('should remove "blur" event listener to form control', () => {
                    const formSpy = jest.spyOn(wrapper.vm, 'onFormControlBlur');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$off');

                    wrapper.vm.remove(inputWrapper.vm);

                    expect(formSpy).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(2, 'blur', expect.any(Function));
                });

                it('should remove validation event listener to form control', () => {
                    const formSpy1 = jest.spyOn(wrapper.vm, 'getFormControlValidationEvents');
                    const formSpy2 = jest.spyOn(wrapper.vm, 'onFormControlValidate');
                    const inputSpy = jest.spyOn(inputWrapper.vm, '$off');

                    wrapper.vm.remove(inputWrapper.vm);

                    expect(formSpy1).toHaveBeenCalled();
                    expect(formSpy2).toHaveBeenCalled();
                    expect(inputSpy).toHaveBeenNthCalledWith(1, 'input', expect.any(Function));
                });
            });

            describe('emitSubmit()', () => {
                beforeEach(() => {
                    const schema = {
                        validate() {},
                        input: {
                            validators: [
                                { rule: 'required' }
                            ]
                        }
                    };

                    wrapper.setProps({
                        value: schema
                    });
                });

                it('should be defined', () => {
                    expect(wrapper.vm.emitSubmit).toBeDefined();
                });

                it('should emit "submit" event', () => {
                    const event = { preventDefault() {} };

                    wrapper.vm.emitSubmit(event);

                    expect(wrapper.emitted().submit).toBeTruthy();
                    expect(wrapper.emitted().submit.length).toBe(1);
                    expect(wrapper.emitted().submit[0]).toEqual([event]);
                });

                it('should validate schema', () => {
                    const spy = jest.spyOn(wrapper.vm.value, 'validate');
                    const event = { preventDefault() {} };

                    wrapper.vm.emitSubmit(event);

                    expect(spy).toHaveBeenCalled();
                });

                it('should return if schema is invalid', () => {
                    const spy = jest.spyOn(wrapper.vm.value, 'validate');
                    const event = { preventDefault() {} };

                    wrapper.vm.value.invalid = true;
                    wrapper.vm.emitSubmit(event);

                    expect(spy).toHaveBeenCalled();
                    expect(wrapper.emitted().submit).not.toBeTruthy();
                });

                it('should emit "submit" if schema is valid', () => {
                    const spy = jest.spyOn(wrapper.vm.value, 'validate');
                    const event = { preventDefault() {} };

                    wrapper.vm.value.input.value = 'value';
                    wrapper.vm.emitSubmit(event);

                    expect(spy).toHaveBeenCalled();
                    expect(wrapper.emitted().submit).toBeTruthy();
                });
            });
        });

        describe('created()', () => {
            it('should add class rules to classes provider', () => {
                const classRulesLength = wrapper.vm.classesProvider.length;

                wrapper.vm.created();
                expect(wrapper.vm.classesProvider.length).toEqual(classRulesLength + 1)
            });

            it('should add "-inline" class if "inline" property is true', () => {
                const rule = wrapper.vm.classesProvider[wrapper.vm.classesProvider.length - 1];

                expect(rule()).toEqual(expect.objectContaining({ '-inline': false }));
                wrapper.setProps({ inline: true });
                expect(rule()).toEqual(expect.objectContaining({ '-inline': true }));
            });
        });
    });
});
