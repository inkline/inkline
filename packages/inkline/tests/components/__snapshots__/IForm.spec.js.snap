// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components IForm should render correctly 1`] = `<form role="form" class="form"></form>`;
